pgmpy/estimators/NOTEARS.py

```diff
@@ -0,0 +1,275 @@
+import numpy as np
+import scipy.linalg as slin
+import scipy.optimize as sopt
+from scipy.special import expit as sigmoid
+from tqdm.auto import trange
+
+from pgmpy import config
+from pgmpy.base import DAG
+from pgmpy.estimators import ExpertKnowledge, StructureEstimator
+from pgmpy.global_vars import logger
+
+
+class NOTEARS(StructureEstimator):
+    """
+    Implementation of the NOTEARS structure learning algorithm.
+
+    NOTEARS is a structure learning algorithm that works by converting the discrete score-based
+    DAG learning problem into a continuous optimisation problem.
+
+    Parameters
+    ----------
+    data: pandas DataFrame object
+        dataframe object where each column represents one variable.
+        (If some values in the data are missing the data cells should be set to `numpy.nan`.
+        Note that pandas converts each column containing `numpy.nan`s to dtype `float`.)
+
+    References
+    ----------
+    Xun Zheng, Bryon Aragam, Pradeep Ravikumar, and Eric P. Xing. 2018. DAGs with NO TEARS: continuous
+    optimization for structure learning. In Proceedings of the 32nd International Conference on Neural
+    Information Processing Systems (NIPS'18). Curran Associates Inc., Red Hook, NY, USA, 9492–9503.
+    """
+
+    def __init__(self, data, **kwargs):
+        super(NOTEARS, self).__init__(data=data, **kwargs)
+
+    def _loss_grad(self, loss_type, X, W):
+        """
+        Calculate the value of the loss function and its gradient.
+
+        Parameters
+        ----------
+        X: numpy.ndarray
+            The data used for DAG estimation.
+
+        W: numpy.ndarray
+            The weighted matrix for the DAG.
+
+        Returns
+        -------
+        loss: float
+            The value of the loss function for given data.
+
+        G_loss: numpy.ndarray
+            The gradient of the loss function for given data.
+        """
+        M = X @ W
+        if loss_type == "l2":
+            R = X - M
+            loss = 0.5 / X.shape[0] * (R**2).sum()
+            G_loss = -1.0 / X.shape[0] * X.T @ R
+
+        elif loss_type == "logistic":
+            loss = 1.0 / X.shape[0] * (np.logaddexp(0, M) - X * M).sum()
+            G_loss = 1.0 / X.shape[0] * X.T @ (sigmoid(M) - X)

```

Comment:
@Nimish-4 Could you also have a look at the codecov warnings and see if you can add tests such that the warning lines get executed when running tests?